{"cells":[{"cell_type":"markdown","metadata":{"id":"yOJlXv__DitX"},"source":["# Ash Ediga\n","\n","## Inspection Introduction\n","\n","Problem 16\n","* searches for specific amino acid sequences within a DNA sequence and its complement by translating DNA codons to amino acids. It utilizes a dictionary, DNA_Codons, mapping DNA codons to amino acids and also creates an inverse dictionary for reverse lookups. A key function, complement_dna, computes the complementary DNA strand, essential for examining both strands of the DNA. Another function, translate_to_amino_acids, translates a DNA sequence into an amino acid sequence using the codon mapping. The script's core, search_amino_acid_sequences, identifies segments in the DNA and its complement that match a given amino acid sequence. Executed from the command line, the script reads DNA and amino acid sequences from standard input and outputs all matching DNA sequence\n","\n","Problem 17\n","* This Python script is designed to calculate the cyclospectrum of a peptide sequence, a task commonly performed in bioinformatics and mass spectrometry analysis. The script defines a dictionary, amino_to_mass, mapping each amino acid to its respective mass. The peptide_mass_calc function computes the total mass of a given peptide by summing the masses of its constituent amino acids. Another function, linear_fragment, generates all possible linear fragments of the peptide, essential for constructing the cyclospectrum. The core function, cyclospectrum, calculates the cyclospectrum by determining the mass of each fragment generated by linear_fragment, including the mass of the whole peptide. The script reads a peptide sequence from standard input and outputs the calculated cyclospectrum, running its main logic in the main function when executed.\n","\n","Problem 18\n","* Cyclopeptide Sequencing Problem, a significant challenge in bioinformatics related to protein structure analysis. The script is organized into three main classes: PeptideKmers, PeptideSpectrum, and CyclopeptideSolver. The PeptideKmers class is responsible for generating all possible subsequences (k-mers) of a given length from a peptide sequence. It takes the peptide and a k-value to produce a list of these k-mers. Following this, the PeptideSpectrum class calculates the mass spectrum of a peptide. It utilizes a predefined dictionary that maps each amino acid to its corresponding mass and computes the mass of every subsequence of the peptide, including the empty sequence and the peptide itself. The resultant mass spectrum is a sorted list of these calculated masses."]},{"cell_type":"markdown","metadata":{"id":"sCVSFm4zDita"},"source":["## Problem16"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"Am1_2XuZDita"},"outputs":[],"source":["import sys\n","\n","# Restructured DNA codons dictionary\n","codon_pairs = [ \n","    \"GCT\": \"A\", \"GCC\": \"A\", \"GCA\": \"A\", \"GCG\": \"A\",\n","    \"TGT\": \"C\", \"TGC\": \"C\",\n","    \"GAT\": \"D\", \"GAC\": \"D\",\n","    \"GAA\": \"E\", \"GAG\": \"E\",\n","    \"TTT\": \"F\", \"TTC\": \"F\",\n","    \"GGT\": \"G\", \"GGC\": \"G\", \"GGA\": \"G\", \"GGG\": \"G\",\n","    \"CAT\": \"H\", \"CAC\": \"H\",\n","    \"ATA\": \"I\", \"ATT\": \"I\", \"ATC\": \"I\",\n","    \"AAA\": \"K\", \"AAG\": \"K\",\n","    \"TTA\": \"L\", \"TTG\": \"L\", \"CTT\": \"L\", \"CTC\": \"L\", \"CTA\": \"L\", \"CTG\": \"L\",\n","    \"ATG\": \"M\",\n","    \"AAT\": \"N\", \"AAC\": \"N\",\n","    \"CCT\": \"P\", \"CCC\": \"P\", \"CCA\": \"P\", \"CCG\": \"P\",\n","    \"CAA\": \"Q\", \"CAG\": \"Q\",\n","    \"CGT\": \"R\", \"CGC\": \"R\", \"CGA\": \"R\", \"CGG\": \"R\", \"AGA\": \"R\", \"AGG\": \"R\",\n","    \"TCT\": \"S\", \"TCC\": \"S\", \"TCA\": \"S\", \"TCG\": \"S\", \"AGT\": \"S\", \"AGC\": \"S\",\n","    \"ACT\": \"T\", \"ACC\": \"T\", \"ACA\": \"T\", \"ACG\": \"T\",\n","    \"GTT\": \"V\", \"GTC\": \"V\", \"GTA\": \"V\", \"GTG\": \"V\",\n","    \"TGG\": \"W\",\n","    \"TAT\": \"Y\", \"TAC\": \"Y\",\n","    \"TAA\": \"_\", \"TAG\": \"_\", \"TGA\": \"_\"\n","]\n","DNA_Codons = dict(codon_pairs)\n","inverse_Codons = {value: key for key, value in DNA_Codons.items()}\n","\n","def complement_dna(sequence):\n","    translation_table = str.maketrans('ACGTN', 'TGCAN')\n","    return sequence[::-1].translate(translation_table)\n","\n","def split_into_codons(dna_sequence, offset=0):\n","    return [dna_sequence[i:i + 3] for i in range(offset, len(dna_sequence), 3)]\n","\n","def translate_to_amino_acids(dna_seq):\n","    return ''.join(DNA_Codons.get(codon, '') for codon in split_into_codons(dna_seq))\n","\n","def search_amino_acid_sequences(dna, amino_acid_seq):\n","    found_sequences = []\n","    dna = dna.upper()\n","    reversed_dna = complement_dna(dna)\n","    amino_acid_seq = amino_acid_seq.upper()\n","    length = len(amino_acid_seq) * 3\n","\n","    for i in range(len(dna) - length + 1):\n","        segment = dna[i:i + length]\n","        if translate_to_amino_acids(segment) == amino_acid_seq:\n","            found_sequences.append(segment)\n","    \n","    for i in range(len(reversed_dna) - length + 1):\n","        segment = reversed_dna[i:i + length]\n","        if translate_to_amino_acids(segment) == amino_acid_seq:\n","            found_sequences.append(complement_dna(segment))\n","\n","    return found_sequences\n","\n","def run_amino_acid_search():\n","    dna_sequence, amino_acid_sequence = (line.strip().upper() for line in sys.stdin)\n","    matching_sequences = search_amino_acid_sequences(dna_sequence, amino_acid_sequence)\n","\n","    for sequence in matching_sequences:\n","        print(sequence)\n","\n","if __name__ == \"__main__\":\n","    run_amino_acid_search()\n"]},{"cell_type":"markdown","metadata":{"id":"VrgboVPpDitb"},"source":["## Problem 17"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"c7_To2hADitb"},"outputs":[],"source":["import sys\n","\n","# Mapping of amino acids to their respective masses\n","amino_to_mass = {\n","    \"G\": 57, \"A\": 71, \"S\": 87, \"P\": 97, \"V\": 99,\n","    \"T\": 101, \"C\": 103, \"I\": 113, \"L\": 113, \"N\": 114,\n","    \"D\": 115, \"K\": 128, \"Q\": 128, \"E\": 129, \"M\": 131,\n","    \"H\": 137, \"F\": 147, \"R\": 156, \"Y\": 163, \"W\": 186\n","}\n","\n","def peptide_mass_calc(peptide):\n","    ''' Calculate the total mass of a peptide '''\n","    return sum(amino_to_mass.get(amino, 0) for amino in peptide)\n","\n","def linear_fragment(peptide):\n","    ''' Generates all linear fragments of the peptide '''\n","    fragments = []\n","    for i in range(1, len(peptide)):\n","        for j in range(len(peptide) - i + 1):\n","            fragments.append(peptide[j:j + i])\n","    return fragments\n","\n","def cyclospectrum(peptide):\n","    ''' Calculate the cyclospectrum of the peptide '''\n","    spectrum = [0]\n","    linear_fragments = linear_fragment(peptide)\n","    \n","    # Add masses of linear fragments\n","    for fragment in linear_fragments:\n","        spectrum.append(peptide_mass_calc(fragment))\n","    \n","    # Add mass of the whole peptide\n","    spectrum.append(peptide_mass_calc(peptide))\n","    \n","    return sorted(spectrum)\n","\n","def main():\n","    # Read the sequence from standard input\n","    sequence = sys.stdin.readline().strip()\n","\n","    # Calculate and print the cyclospectrum\n","    spectrum = cyclospectrum(sequence)\n","    print(' '.join(map(str, spectrum)))\n","\n","if __name__ == \"__main__\":\n","    main()\n"]},{"cell_type":"markdown","metadata":{"id":"JVMooyHDDitc"},"source":["## Problem 18"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"ZjWFmsbhDitc"},"outputs":[],"source":["import sys\n","\n","class PeptideKmers:\n","    ''' Generate all possible k-mers of a given length from a peptide sequence '''\n","    def __init__(self, peptide, k):\n","        self.peptide = peptide\n","        self.k = k\n","\n","    def generate_kmers(self):\n","        return [self.peptide[i:i+self.k] for i in range(len(self.peptide) - self.k + 1)]\n","\n","class PeptideSpectrum:\n","    ''' Calculate the mass spectrum of a peptide '''\n","    mass_of_amino_acids = {\n","        'A': 71, 'G': 57, 'M': 131, 'S': 87, 'C': 103,\n","        'H': 137, 'N': 114, 'T': 101, 'D': 115, 'I': 113,\n","        'P': 97, 'V': 99, 'E': 129, 'K': 128, 'Q': 128,\n","        'W': 186, 'F': 147, 'L': 113, 'R': 156, 'Y': 163\n","    }\n","\n","    def __init__(self, peptide):\n","        self.peptide = peptide\n","\n","    def calculate_spectrum(self):\n","        subsets = [''] + [sub for k in range(1, len(self.peptide)+1) for sub in PeptideKmers(self.peptide, k).generate_kmers()]\n","        return sorted([sum(self.mass_of_amino_acids.get(aa, 0) for aa in subset) for subset in subsets])\n","\n","class CyclopeptideSolver:\n","    ''' Solve the Cyclopeptide Sequencing Problem given a spectrum '''\n","    def __init__(self, spectrum):\n","        self.spectrum = spectrum\n","\n","    def find_sequences(self):\n","        active_peptides = [[]]\n","        final_sequences = []\n","\n","        while active_peptides:\n","            extended_peptides = [pep + [mass] for pep in active_peptides for mass in set(PeptideSpectrum.mass_of_amino_acids.values())]\n","            active_peptides = []\n","\n","            for peptide in extended_peptides:\n","                if sum(peptide) == max(self.spectrum):\n","                    if PeptideSpectrum(peptide).calculate_spectrum() == self.spectrum:\n","                        final_sequences.append(peptide)\n","                elif all(mass in self.spectrum for mass in PeptideSpectrum(peptide).calculate_spectrum()):\n","                    active_peptides.append(peptide)\n","\n","        return final_sequences\n","\n","def main():\n","    input_spectrum = [int(mass) for mass in sys.stdin.readline().strip().split()]\n","    solver = CyclopeptideSolver(input_spectrum)\n","    for sequence in solver.find_sequences():\n","        print('-'.join(map(str, sequence)), end=' ')\n","\n","if __name__ == \"__main__\":\n","    main()\n"]},{"cell_type":"markdown","metadata":{"id":"2AJt7r1aDlah"},"source":["## Inspection Results"]},{"cell_type":"markdown","metadata":{},"source":["Konstantinos: There is no error handling in your code. you're generating all possible k-mers for each k value. This is a good approach, but it could be more efficient if large peptides are processed, as it generates a lot of intermediate lists. \n","\n","Trevor: In CyclopeptideSolver.find_sequences, the nested loops with list comprehensions might become inefficient for larger spectra due to the creation of numerous temporary lists. You might want to profile this part if you're dealing with large datasets.\n","\n"]}],"metadata":{"colab":{"provenance":[]},"kernelspec":{"display_name":"Python 3.11.2 64-bit","language":"python","name":"python3"},"language_info":{"name":"python","version":"3.11.2"},"vscode":{"interpreter":{"hash":"aee8b7b246df8f9039afb4144a1f6fd8d2ca17a180786b69acc140d282b71a49"}}},"nbformat":4,"nbformat_minor":0}
